boards: stm32l562e_dk: add support for SPI

yestin opened the pull request:
Add spi1 interface for stm32l552xx and stm32l562xx microcontrollers
and enable spi1 that connects to STM module SPBTLE-RFTR on the
stm32l562e_dk board.

Tested the configuration with st_ble_sensor sample + ST BLE Sensor
app on Android phone.

Signed-off-by: Yestin Sun <sunyi0804@gmail.com>

erwango commented on boards/arm/stm32l562e_dk/Kconfig.defconfig:
```diff
@@ -8,4 +8,27 @@ if BOARD_STM32L562E_DK
+# Disable Flow control
+config BT_HCI_ACL_FLOW_CONTROL
+	default n
+config BT_HCI_VS_EXT
```
nit: break one line here.

yestin replied:
Done

erwango reviewed (changes requested):
Thanks for this addition!

One minor comment on the code.
Otherwise, please split the pull request in two:
- dts: Add spi to L5, ...
- boards: Enable SPI and BLE on stm32l5...

Btw, thanks for your interest on this board. I noticed 2 things when testing the ble sample:
- LED status is opposite on the board and on the app. Do you see that as well ? If yes, is it due to LED dts definition on this board?
- I'm having trouble flashing the board using pyocd, each time I want to reflash it, I need to perform a full chip erase using STM32Cube programmer. Is that the same on your sde. If yes, I wonder if we should move default runner to STM32Cube programmer (in board.cmake file) :
```
board_runner_args(stm32cubeprogrammer "--port=swd" "--reset=hw")
include(${ZEPHYR_BASE}/boards/common/stm32cubeprogrammer.board.cmake)
```

yestin committed:
soc: arm: st: Add spi to stm32l5 MCUs

Add spi1 interface for stm32l552xx and stm32l562xx microcontrollers

Signed-off-by: Yestin Sun <sunyi0804@gmail.com>

yestin commented:
> Otherwise, please split the pull request in two:
> 
> * dts: Add spi to L5, ...
> * boards: Enable SPI and BLE on stm32l5...

I have split the patch into two for dts and board changes respectively.

> * LED status is opposite on the board and on the app. Do you see that as well ? If yes, is it due to LED dts definition on this board?

Thanks for catching this. You are right. It turns out I made a mistake in the LED dts configurations. Both LEDs shall be active low instead of active high. The change has been made in the new patches.

> * I'm having trouble flashing the board using pyocd, each time I want to reflash it, I need to perform a full chip erase using STM32Cube programmer. Is that the same on your sde. If yes, I wonder if we should move default runner to STM32Cube programmer (in board.cmake file) :

I do not see this specific issue. My development environment is MacOSX and I cloned the openocd source from https://github.com/zephyrproject-rtos/openocd/tree/zephyr-20200722 @c41269f3f81e85de24b1ee63e99b89abaefbae41 and built it locally, which could matter.

erwango commented:
@yestin txs for the changes.

> I do not see this specific issue. My development environment is MacOSX and I cloned the openocd source from https://github.com/zephyrproject-rtos/openocd/tree/zephyr-20200722 @c41269f3f81e85de24b1ee63e99b89abaefbae41 and built it locally, which could matter.

Well, it seems there is a desynchro between you're local environment and what other users can enjoy.

As well as pyocd include being first (ie default), the following lines in `board.cmake`, are forcing the use of pyocd:
```
set_ifndef(BOARD_DEBUG_RUNNER pyocd)
set_ifndef(BOARD_FLASH_RUNNER pyocd)
```
Then, if trying to use openocd, this fails with following error:
```
Error: Debug Adapter has to be specified, see "adapter driver" command
```
And indeed, to use openocd, we a `board/arm/foo/support/openocd.cfg` to provide openocd configuration (which I guess you have locally). You can check other boards for more info.
In order to help others using this work, it would be nice to enable a working way of flashing upstream. Openocd would be ideal (enables debug), but STM32CubeProgrammer works as well.
Would you have some time to check this ?

erwango commented on boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi:
```diff
@@ -12,11 +12,11 @@
 		red_led_9: led_9 {
-			gpios = <&gpiod 3 GPIO_ACTIVE_HIGH>;
+			gpios = <&gpiod 3 GPIO_ACTIVE_LOW>;
 			label = "User LD9";
 		};
 		green_led_10: led_10 {
-			gpios = <&gpiog 12 GPIO_ACTIVE_HIGH>;
+			gpios = <&gpiog 12 GPIO_ACTIVE_LOW>;
```
This would theoretically require a dedicated commit.
At least it would be required to mention this in the commit message

yestin replied:
Done. Moved the GPIO configuration fix into a separate commit.

erwango reviewed (changes requested):
One last nit and we're good to go!

yestin committed:
boards: stm32l562e_dk: Fix inverted GPIO flags for user LEDs

Fix the GPIO configurations for both user LEDs on the stm32l562e_dk
board that shall be active low.

Signed-off-by: Yestin Sun <sunyi0804@gmail.com>

yestin committed:
boards: stm32l562e_dk: Enable SPI and BLE

Enable spi1 interface that connects to STM module SPBTLE-RFTR on the
stm32l562e_dk board.

Tested the configuration with st_ble_sensor sample + ST BLE Sensor
app on Android phone.

Signed-off-by: Yestin Sun <sunyi0804@gmail.com>

yestin commented:
> And indeed, to use openocd, we a `board/arm/foo/support/openocd.cfg` to provide openocd configuration (which I guess you have locally). You can check other boards for more info.
> In order to help others using this work, it would be nice to enable a working way of flashing upstream. Openocd would be ideal (enables debug), but STM32CubeProgrammer works as well.
> Would you have some time to check this ?

To be honest, I started the initial stm32l562e_dk board bringup based on nucleo_l552ze_q board since I am not familiar with openocd. I replaced `target=stm32l552zetxq` with `target=stm32l562qeixq` and it was working, so I did not spend much time on it. I checked my local environment but I don't have `board/arm/foo/support/openocd.cfg` for this specific board.

But sure, I can investigate the discrepancy. If you don't mind, I can take my time to investigate and submit a new PR for this specific topic. Also would you mind letting me know what your environment is and the Zephyr SDK version you are using so that I can set up similar environment to reproduce the issue.

erwango commented:
> To be honest, I started the initial stm32l562e_dk board bringup based on nucleo_l552ze_q board since I am not familiar with openocd. I replaced target=stm32l552zetxq with target=stm32l562qeixq and it was working, so I did not spend much time on it. I checked my local environment but I don't have board/arm/foo/support/openocd.cfg for this specific board.

Ok, indeed I'm facing same issue on  nucleo_l552ze_q. 

> But sure, I can investigate the discrepancy. If you don't mind, I can take my time to investigate and submit a new PR for this specific topic. Also would you mind letting me know what your environment is and the Zephyr SDK version you are using so that I can set up similar environment to reproduce the issue.

Would be nice indeed if you have time to share your openocd config. I'm using Ubuntu 18.04 and latest SDK version (0.12 beta).
This is a totally different topic as current PR so indeed this is not at all blocking this one.

yestin commented:
> Would be nice indeed if you have time to share your openocd config. I'm using Ubuntu 18.04 and latest SDK version (0.12 beta).
> This is a totally different topic as current PR so indeed this is not at all blocking this one.

Ok. I am able to see the flashing error after I set up the Ubuntu 18.04 environment.

After some digging on my local MacOSX environment, I find that I have the STM32L5xx_DFP pack installed for pyocd. I must have installed it manually when the flashing was not working.

```$ pyocd pack --find stm32l562qeixq
  Part             Vendor   Pack            Version   Installed
-----------------------------------------------------------------
  STM32L562QEIxQ   Keil     STM32L5xx_DFP   1.2.0     True```

That's why my flashing works with pyocd. So I don't have a special openocd config since I use pyocd to flash on MacOSX.

I will try to get the flashing working with openocd. I notice on Linux, the openocd config files are part of the Zephyr SDK, so I assume I will need to add the new target config for stm32l5x and a new board config for stm32l562e_dk in this repository https://github.com/zephyrproject-rtos/openocd. Please let me know if I am heading for the right direction.

If everything goes well, I can add a new board config for nucleo_l552ze_q as well since they shall be the same; or I will ask for help :).

erwango commented:
> So I don't have a special openocd config since I use pyocd to flash on MacOSX.

Ok. So I get it better.
Btw, I suspect that L5 is not supported at all on openocd. So if you don't find clue that it is supported, no need to loose much time on this, as adding it could be somewhat more complex than other STM32 series.

> That's why my flashing works with pyocd.

On my side, 1.3.0 is installed and yet I'm finding issues (Need to fully erase the flash if I want to flash a new application and the board has been unplugged).
```
$ pyocd pack --find stm32l562qeixq
  Part             Vendor   Pack            Version   Installed  
-----------------------------------------------------------------
  STM32L562QEIxQ   Keil     STM32L5xx_DFP   1.3.0     True  
```

yestin commented:
> Btw, I suspect that L5 is not supported at all on openocd. So if you don't find clue that it is supported, no need to loose much time on this, as adding it could be somewhat more complex than other STM32 series.

Looks like the stm32l5x support for openocd is being actively worked on http://openocd.zylin.com/#/c/5510/9. I am looking forward to the v0.11.0 release.

ABOSTM commented on boards/arm/stm32l562e_dk/doc/index.rst:
```diff
@@ -184,6 +186,7 @@ Default Zephyr Peripheral Mapping:
 
 - USART_1 TX/RX : PA9/PA10
 - I2C_1 SCL/SDA : PB6/PB7
+- SPI_1 SCL/SDA : PB6/PB7 (BT SPI bus)
```
Wrong SPI signals and pins

yestin replied:
Thanks for catching the typo. The new patch has been uploaded to fix this.

ABOSTM reviewed (approved):
LGTM
